cylc/flow/scripts/config.py

```diff
@@ -110,6 +111,16 @@ def get_option_parser() -> COP:
             "overrides any settings it shares with those higher up."),
         action="store_true", default=False, dest="print_hierarchy")
 
+    parser.add_option(
+        '--json',
+        help=(
+            'Print metadata from a Cylc configuration.'
+        ),
+        default=False,
+        action='store_true',
+        dest='json'
+    )
+
```

Comment:
The `cylc config` command can be used in multiple ways:

* View workflow/global config.
* View config with/without defaults.
* View only selected sections/settings or view all.
* View in "oneline" mode.
* View platforms/platform names.

This new `--json` option will ignore all other options (eng. `--defaults`, `-i`, `--one-line`, etc) which is a tad confusing.

I would suggest adding a `jdump` (i.e. json dump) method to the ParsecConfig object (to complement the `idump` method). That way the sparse/dense options will work as expected. Then raising an `InputError` if any conflicting CLI options are specified (and updating the `--json` help text to reflect this).

Reply:
I've done this, but I think that it'll want a couple more tests.

Reply:
> Although I am not sure why --json can't just be used to control the output format as is currently done

I'm a little confused about what you mean by this comment - that's what I think `--json` _is_ doing.

Reply:
Yes, I don't see a problem with it. But in Oliver's comment it seems there was a different idea about what `--json` should be doing.

Reply:
The question is: should there be any mutually exclusive options to `--json`?

Reply:
Tested:
- `--one-line` returns un-prettified JSON at the moment; seems reasonable to me, but different to how it behaves for non-json output (though I don't understand what that is meant to be doing in those cases)
- `--print-hierearchy`, `--platform-names`, `--platforms` - not affected by `--json`; I don't see a reason why these couldn't be outputted in JSON, but do we have a need?

Reply:
> Or was Oliver talking about an original implementation that did something different, that you have since changed?

Yes.